File: Ollama-Flask/controllers/embed_controller.py
from flask import request, jsonify
import urllib.request
import urllib.error
import json

def generate_embeddings():
    try:
        data = request.get_json()
        if not data:
            return jsonify({"error": "Invalid JSON payload"}), 400

        texts = data.get("texts")
        model = data.get("model", "nomic-embed-text")

        if not texts or not isinstance(texts, list):
            return jsonify({"error": "Please provide an array of strings in the \"texts\" field."}), 400

        if len(texts) > 100:
            return jsonify({"error": "Exceeded maximum limit of 100 texts per request."}), 400

        if len(texts) == 0:
            return jsonify({"error": "The texts array cannot be empty."}), 400

        req_data = json.dumps({
            "model": model,
            "input": texts
        }).encode("utf-8")

        req = urllib.request.Request(
            "http://127.0.0.1:11434/api/embed",
            data=req_data,
            headers={"Content-Type": "application/json"},
            method="POST"
        )

        try:
            with urllib.request.urlopen(req) as response:
                response_body = response.read().decode("utf-8")
                response_json = json.loads(response_body)

                return jsonify({
                    "success": True,
                    "model": response_json.get("model"),
                    "embeddings": response_json.get("embeddings")
                }), 200
        except urllib.error.HTTPError as e:
            error_data = e.read().decode("utf-8")
            return jsonify({"error": f"Ollama API error: {error_data}"}), 500
        except urllib.error.URLError as e:
            return jsonify({"error": f"Failed to connect to Ollama API: {e.reason}"}), 500

    except Exception as e:
        print(f"Embedding Generation Error: {e}")
        return jsonify({"error": "Internal Server Error while generating embeddings."}), 500

File: Ollama-NodeJS/controllers/embed_controller.mjs
export const generateEmbeddings = async (req, res) => {
    try {
        const { texts, model = "nomic-embed-text" } = req.body;

        if (!texts || !Array.isArray(texts)) {
            return res.status(400).json({ error: "Please provide an array of strings in the \"texts\" field." });
        }

        if (texts.length > 100) {
            return res.status(400).json({ error: "Exceeded maximum limit of 100 texts per request." });
        }

        if (texts.length === 0) {
            return res.status(400).json({ error: "The texts array cannot be empty." });
        }

        // Calling Ollama's native /api/embed endpoint
        const response = await fetch("http://127.0.0.1:11434/api/embed", {
            method: "POST",
            headers: {
                "Content-Type": "application/json"
            },
            body: JSON.stringify({
                model: model,
                input: texts
            })
        });

        if (!response.ok) {
            const errorData = await response.text();
            throw new Error(`Ollama API error: ${errorData}`);
        }

        const data = await response.json();

        return res.status(200).json({
            success: true,
            model: data.model,
            embeddings: data.embeddings
        });

    } catch (error) {
        console.error("Embedding Generation Error:", error);
        return res.status(500).json({ error: "Internal Server Error while generating embeddings." });
    }
};
